materialize-webhook/driver.go

```diff
@@ -161,87 +197,107 @@ func (d *transactor) Load(it *m.LoadIterator, _ func(int, json.RawMessage) error
-		} else {
-			b.WriteString("[\n")
-		}
-		if _, err := b.Write(it.RawJSON); err != nil {
-			return nil, err
-		}
-	}
-
-	for i := range bodies {
-		bodies[i].WriteString("\n]")
-	}
-
-	for i, address := range d.addresses {
-		var address = address.String()
-		var body = &bodies[i]
-
-		for attempt := 0; true; attempt++ {
-			select {
-			case <-ctx.Done():
-				return nil, ctx.Err()
-			case <-time.After(backoff(attempt)):
-				// Fallthrough.
-			}
-
-			request, err := http.NewRequest("POST", address, bytes.NewReader(body.Bytes()))
+		group.Go(func() error {
+			request, err := http.NewRequest("POST", address, pipeReader)
 			if err != nil {
-				return nil, fmt.Errorf("http.NewRequest(%s): %w", address, err)
+				return fmt.Errorf("http.NewRequest(%s): %w", address, err)
 			}
+
 			request.Header.Add("Content-Type", "application/json")
 
+			for i := range d.customHeaders {
+				header := d.customHeaders[i]
+				request.Header.Add(header.Name, header.Value)
+			}
+
 			response, err := http.DefaultClient.Do(request)
+
 			if err == nil {
 				err = response.Body.Close()
 			}
+
 			if err == nil && (response.StatusCode < 200 || response.StatusCode >= 300) {
-				err = fmt.Errorf("unexpected webhook response code %d from %s",
+				return fmt.Errorf("unexpected webhook response code %d from %s",
 					response.StatusCode, address)
 			}
 
-			if err == nil {
-				body.Reset() // Reset for next use.
-				break
-			} else if attempt == 10 {
-				return nil, fmt.Errorf("webhook failed after many attempts: %w", err)
+			return nil
+		})
+
+		pipeWriter = w
+	}
+
+	finishWebhook := func() error {
+		if pipeWriter == nil {
+			return nil
+		}
+
+		pipeWriter.Write([]byte("]"))
+		pipeWriter.Close()
+
+		if err := group.Wait(); err != nil {
+			return fmt.Errorf("group.Wait(): %w", err)
+		}
+
+		pipeWriter = nil
+		return nil
+	}
+
+	const requestSizeCutoff = 1024 * 1024 // 100 MiB
+	byteCount := 0
+	var previousAddress string
+
+	for it.Next() {
+		address := d.addresses[it.Binding].String()
+
+		// The webhook for the previous binding must finish before the next binding's webhook starts.
+		if previousAddress != "" && address != previousAddress {
+			if err := finishWebhook(); err != nil {
+				return nil, fmt.Errorf("finishWebhooks: %w", err)
+			}
+		}
+
+		previousAddress = address
+
+		if pipeWriter == nil {
+			startWebhook(address)
+			pipeWriter.Write([]byte("["))
+		} else {
+			pipeWriter.Write([]byte(","))
+		}
+
+		pipeWriter.Write(it.RawJSON)
+		byteCount += len(it.RawJSON)
+
+		if byteCount >= requestSizeCutoff {
+			if err := finishWebhook(); err != nil {
+				return nil, fmt.Errorf("finishWebhook: %w", err)
 			}
 
-			log.WithFields(log.Fields{
-				"err":     err,
-				"attempt": attempt,
-				"address": address,
-			}).Error("failed to invoke Webhook (will retry)")
+			byteCount = 0
 		}
 	}
+
+	if err := finishWebhook(); err != nil {
+		return nil, fmt.Errorf("finishWebhook: %w", err)
+	}
+
+	if err := it.Err(); err != nil {
```

Comment:
Probably would be good to move this error check to right after the iterator loop, which is a more typical pattern.